libs/libcommon/src/libcommon/utils.py

```diff
@@ -148,3 +155,68 @@ def raise_if_blocked(
                 "This dataset has been disabled for now. Please open an issue in"
                 " https://github.com/huggingface/datasets-server if you want this dataset to be supported."
             )
+
+
+FuncT = TypeVar("FuncT", bound=Callable[..., Any])
+RETRY_SLEEPS = (1, 1, 1, 10, 10, 10, 60, 60, 60, 10 * 60)
+RETRY_ON: tuple[type[Exception]] = (Exception,)
+
+
+class retry:
+    """retries with an increasing sleep before every attempt"""
+
+    def __init__(self, sleeps: Sequence[float] = RETRY_SLEEPS, on: Sequence[type[Exception]] = RETRY_ON) -> None:
+        self.sleeps = sleeps
+        self.on = on
+
+    def __call__(self, func: FuncT) -> FuncT:
+        @functools.wraps(func)
+        def decorator(*args: Any, **kwargs: Any) -> Any:
+            attempt = 0
+            last_err = None
+            while attempt < len(self.sleeps):
+                try:
+                    """always sleep before calling the function. It will prevent rate limiting in the first place"""
+                    duration = self.sleeps[attempt]
+                    logging.info(f"Sleep during {duration} seconds to preventively mitigate rate limiting.")
+                    time.sleep(duration)
+                    return func(*args, **kwargs)
+                except tuple(self.on) as err:
+                    logging.info(f"Got a {type(err)}. Let's retry.")
+                    last_err = err
+                    attempt += 1
+            raise RuntimeError(f"Give up after {attempt} attempts. The last one raised {type(last_err)}") from last_err
+
+        return cast(FuncT, decorator)
+
+
+HF_HUB_HTTP_ERROR_RETRY_SLEEPS = [1, 1, 1, 10, 10, 10]
+
+
+def download_file_from_hub(
+    repo_type: str,
+    revision: str,
+    repo_id: str,
+    filename: str,
+    local_dir: Union[str, Path],
+    hf_token: Optional[str],
+    cache_dir: Union[str, Path, None] = None,
+    force_download: bool = False,
+    resume_download: bool = False,
+) -> None:
+    # Force hf_transfer usage
+    constants.HF_HUB_ENABLE_HF_TRANSFER = True
```

Comment:
I don't think this will work because (unfortunately) we don't do `from huggingface_hub import constants` but `from huggingface.constants import HF_HUB_ENABLE_HF_TRANSFER` in hfh modules (related issue https://github.com/huggingface/huggingface_hub/issues/1172 - never did it as low priority).

But in general, why do you want to hardcode this value here? Wouldn't it be easier to set the environment variable on the machine where `dataset-server` is run? Even a `HF_HUB_ENABLE_HF_TRANSFER=1 python my_script.py` would be enough (no need to set it globally). I'm a bit reluctant with hardcoding this because as said to @lhoestq yesterday, hf_transfer is not always better. On a smaller machine or a machine with an average connection, you won't have a speed increase and even something a speed decrease. I guess it's not a problem on the dataset-server machines but for local debugging that can be annoying/counter-productive.

Reply:
If you really want to patch `HF_HUB_ENABLE_HF_TRANSFER`, I would do like this. The advantage of `mock.patch.object` is that the value is reset to its original value when exiting the context. Just tried it locally and confirm it works (also confirmed locally `constants.HF_HUB_ENABLE_HF_TRANSFER = True` doesn't :confused:)

```py
from unittest.mock import patch
from huggingface_hub import file_download, hf_hub_download

with patch.object(file_download, "HF_HUB_ENABLE_HF_TRANSFER", True):
    hf_hub_download("gpt2", "model.safetensors")
```